src/client/common/utils/version.ts

```diff
@@ -3,7 +3,290 @@
+ * is not set.  Lower properties will not be set if a higher property
+ * is not.
+ *
+ * Note that any object can be forced to look like a VersionInfo and
+ * any of the properties may be forced to hold a non-number value.
+ * To resolve this situation, pass the object through
+ * `normalizeVersionInfo()` and then `validateVersionInfo()`.
+ */
+export type BasicVersionInfo = {
+    major: number;
+    minor: number;
+    micro: number;
+    // There is also a hidden `unnormalized` property.
+};
+
+function normalizeVersionPart(part: number): number {
+    // Any -1 values where the original is not a number are handled in validation.
+    if (typeof part === 'number') {
+        if (isNaN(part) || part < 0) {
+            // We leave this as a marker.
+            return -1;
+        }
+        return part;
+    }
+    if (typeof part === 'string') {
+        const parsed = parseInt((part as unknown) as string, 10);
+        if (isNaN(parsed) || parsed < 0) {
+            return -1;
+        }
+        return parsed;
+    }
+    if (part === undefined || part === null) {
+        return -1;
+    }
+    return -1;
+}
+
+type RawBasicVersionInfo = BasicVersionInfo & {
+    unnormalized?: {
+        // tslint:disable:no-any
+        major?: any;
+        minor?: any;
+        micro?: any;
+        // tslint:enable:no-any
+    };
+};
+
+/**
+ * Make a copy and set all the properties properly.
+ */
+export function normalizeBasicVersionInfo<T extends BasicVersionInfo>(info: T): T {
+    if (!info) {
+        const empty = { major: -1, minor: -1, micro: -1 };
+        ((empty as unknown) as RawBasicVersionInfo).unnormalized = {
+            major: undefined,
+            minor: undefined,
+            micro: undefined
+        };
+        return empty as T;
+    }
+    const norm: T = { ...info };
+    const raw = (norm as unknown) as RawBasicVersionInfo;
+    if (raw.unnormalized === undefined) {
+        raw.unnormalized = { ...norm };
+        norm.major = normalizeVersionPart(norm.major);
+        norm.minor = normalizeVersionPart(norm.minor);
+        norm.micro = normalizeVersionPart(norm.micro);
+    }
+    return norm;
+}
+
+function validateVersionPart(prop: string, part: number, unnormalized?: unknown) {
+    if (typeof part !== 'number' || isNaN(part)) {
+        throw Error(`invalid ${prop} version (not normalized)`);
+    }
+    if (part === 0 || part > 0) {
+        return;
+    }
+    if (typeof unnormalized === 'number') {
+        return;
+    }
+    if (unnormalized === undefined || unnormalized === null) {
+        return;
+    }
+    if (typeof unnormalized === 'string' && !isNaN(parseInt(unnormalized, 10))) {
+        return;
+    }
+    throw Error(`invalid ${prop} version (failed to normalize)`);
+}
+
+/**
+ * Fail if any properties are not set properly.
+ *
+ * The info is expected to be normalized already.
+ */
+export function validateBasicVersionInfo<T extends BasicVersionInfo>(info: T) {
```

Comment:
```suggestion
export function validateBasicVersionInfo<T extends BasicVersionInfo>(info?: T) {
```

<img width="582" alt="image" src="https://user-images.githubusercontent.com/51720070/91366264-1de88d80-e7b8-11ea-8b75-c3adb2cd9960.png">

Reply:
As with `normalizeBasicVersionInfo()`, the declared type sets the expectation, even if internally we are treating the provided data defensively.

Reply:
Resolving this conversation since it boils down to https://github.com/microsoft/vscode-python/pull/13591#discussion_r478741817